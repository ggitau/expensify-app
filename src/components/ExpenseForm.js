import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
constructor(props){
  super(props);
  this.state = {
    description: props.expense ? props.expense.description:'',
    note: props.expense?props.expense.note:'',
    amount:props.expense?(props.expense.amount/100).toString():'',
    createdAt: props.expense?moment(props.expense.createdAt):moment(),
    calendarFocused: false,
    error: ''
  };
}

  onDescriptionChange = (e) => {
    const description=e.target.value;
    this.setState(()=>({ description }));
  };

  onNoteChange = (e) => {
    const note=e.target.value;
    this.setState(()=>({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(()=>({amount}));
    }
  };

  onCreatedAtChange = (createdAt) => {
    if (createdAt) {
        this.setState(()=>({ createdAt }));
    }
  };

  onFocusChange = ({focused}) => {
    this.setState(()=>({calendarFocused:focused}));
  };

  onSubmit = (e) => {
    e.preventDefault();
    let error = '';
    if (!this.state.description || !this.state.amount) {
      error='Please provide description and amount';
    }else{
      error='';
      this.props.onSubmit({
        description:this.state.description,
        amount:parseFloat(this.state.amount,10)*100,
        createdAt:this.state.createdAt.valueOf(),
        note:this.state.note
      })
    }
    this.setState(()=>({error}));

  };

  render(){
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}/>
            <input
              type="text"
              value={this.state.amount}
              onChange={this.onAmountChange}
              placeholder="Amount"/>
              <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onCreatedAtChange}
                focused={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={(day) => false }/>
              <textarea
                onChange={this.onNoteChange}
                value={this.state.note}
                placeholder="Add a note for your expense(Optional)">
              </textarea>
              <button>Add Expense</button>
            </form>
          </div>
        );
    }
}
