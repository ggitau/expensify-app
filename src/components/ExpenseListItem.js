import React from 'react';
import {removeExpense} from '../actions/expenses';
import {Link} from 'react-router-dom';
const ExpenseListItem = ({id,description,amount,createdAt}) => (
    <div>
    <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
    <p>{amount}</p>
    <p>{createdAt}</p>

    </div>
);
export default ExpenseListItem;
