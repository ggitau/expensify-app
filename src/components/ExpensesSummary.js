import React from 'react';
import numeral from 'numeral';
import {connect} from 'react-redux';
import totalExpenses from '../selectors/expenses-total';
import visibleExpenses from '../selectors/expenses';

export class ExpensesSummary extends React.Component{
    render(){
        return (
            <div>
                <p>
                Viewing {this.props.expensesCount}
                {
                    this.props.expensesCount>1 || this.props.expensesCount == 0
                    ? ' expenses '
                    :' expense '
                }
                totalling {this.props.expensesTotal}
                </p>
            </div>
        );
    }
}
const mapStateToProps = (state) => {
    return {
        expensesTotal:numeral(totalExpenses(visibleExpenses(state.expenses,state.filters))/100).format('$0,0.00'),
        expensesCount:visibleExpenses(state.expenses,state.filters).length
    };
};
export default connect(mapStateToProps)(ExpensesSummary);