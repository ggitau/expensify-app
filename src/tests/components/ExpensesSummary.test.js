import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('should correctly render ExpensesSummary when expenses count > 1 ', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal="$19.23" expensesCount={2}/>)
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary when expenses count = 1 ', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal="$19.23" expensesCount={1}/>)
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary when expenses count = 0 ', () => {
    const wrapper = shallow(<ExpensesSummary expensesTotal="$19.23" expensesCount={0}/>)
    expect(wrapper).toMatchSnapshot();
});