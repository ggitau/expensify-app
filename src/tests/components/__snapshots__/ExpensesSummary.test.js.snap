// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary when expenses count = 0  1`] = `
<div>
  <p>
    Viewing 
    0
     expenses 
    totalling 
    $19.23
  </p>
</div>
`;

exports[`should correctly render ExpensesSummary when expenses count = 1  1`] = `
<div>
  <p>
    Viewing 
    1
     expense 
    totalling 
    $19.23
  </p>
</div>
`;

exports[`should correctly render ExpensesSummary when expenses count > 1  1`] = `
<div>
  <p>
    Viewing 
    2
     expenses 
    totalling 
    $19.23
  </p>
</div>
`;
